tensorflow/core/platform/default/logging.cc

```diff
@@ -31,28 +31,41 @@ limitations under the License.
 #include <string.h>
 #include <time.h>
 
-#include <string>
+#include <algorithm>
+#include <mutex>
+#include <queue>
 #include <unordered_map>
+#include <vector>
 
 namespace tensorflow {
+namespace internal {
 
-void TFAddLogSink(TFLogSink* sink) {
-  // LogSink is not implemented.
-  // If necessary, one can add the log sink support as follows.
-  // 1. Define a global vector<TFLogSink> to keep track of all registered
-  //    TFLogSink objects. Protect the global vector with mutex to make it
-  //    thread-safe.
-  // 2. Add/remove elements from the global vector<TFLogSink> in TFAddLogSink
-  //    and TFRemoveLogSink function
-  // 3. Add logic in LogMessage::GenerateLogMessage() below to dispatch log
-  //    messages to all the registered log sinks.
-}
+class TFLogSinks {
+  public:
+    static TFLogSinks& Instance();
 
-void TFRemoveLogSink(TFLogSink* sink) {
-  // LogSink is not implemented.
-}
+    void Add(TFLogSink* sink);
+    void Remove(TFLogSink* sink);
+    void Send(const TFLogEntry& entry);
 
-namespace internal {
+  private:
+    TFLogSinks() = default;
+    std::vector<std::unique_ptr<TFLogSink>>::iterator FindSink(TFLogSink* sink);
+    void SendToSink(TFLogSink& sink, const TFLogEntry& entry);
+
+#ifndef NO_DEFAULT_LOGGER
+    TFDefaultLogSink default_;
+#else
+    std::queue<TFLogEntry> queue_;
+    static const size_t MAX_QUEUE = 128;
+#endif // NO_DEFAULT_LOGGER
+
+    static std::mutex instanceMutex_;
+    static std::atomic<TFLogSinks*> instance_;
+    std::mutex mutex_;
+    std::vector<std::unique_ptr<TFLogSink>> sinks_;
```

Comment:
Adding and removing API takes a raw pointer, e.g., `TFAddLogSink(TFLogSink* sink)` which implies it doesn't take the ownership.  So let's use `std::vector<TFLogSink*>`.

Reply:
The log sink API must take ownership. It's simply too unsafe otherwise. What is your convention for taking ownership?

Reply:
To elaborate on my decision to use a raw pointer here: the reason I chose to do it this way is because I use the pointer value as the logging sink ID. I wanted the user to be able to retain the pointer value for future use, while the logger would take ownership.

These were my considerations for each case:
* If I used `std::unique_ptr`, the user would have to jump through some hoops to retain the pointer value. 
* We could use a `std::shared_ptr`, which is probably the safest, but shared ownership is not ideal in this case IMHO, though my reasons are mostly academic
* I could actually make your suggested change, because it really doesn't matter whether the pointers are freed or not: their lifetime will be the same as application lifetime. The big problem here is if we give user the ownership, they can really bring the application down in unexpected ways by freeing that pointer.

I'm going to wait on your answer before doing anything on this one, I just wanted you to know the "why". I feel that I made the "least bad" choice here, but I know that your coding standards are drastically different than mine, so my choice may not be the right one for the situation.

Reply:
OK, that's what we use as well. Your `TFGetLogSinks()` suggestion alleviates the issue with not having a way to remove an existing log sink. 

The reason I'm so adamant about the user not having the sole ownership of a log sink is because this allows a situation where a user will free a log sink while there's still logging to be done. Usually this happens during a program shutdown and can cause behavior that is difficult to debug. We use `shared_ptr` for our logging, but `unique_ptr` is perfectly acceptable as well.

Your comment about just having a static log sink is not always feasible: there are use cases where a user would want to dynamically add and remove log sinks during execution. In that case, I'd rather the user control the log sink lifetime by actively removing the log sink when it's no longer needed than to have a situation where log sink object has been removed, but the log sink is still registered.

Having said that, it's your library, so I don't have a problem with keeping the `TFLogSink*` signature, but not taking ownership. My preference would be to change the signature to `std::unique_ptr<TFLogSink>` or `std::shared_ptr<TFLogSink>`. Please let me know which way you'd like me to go.

Reply:
Let's keep `TFLogSink*` signature, but please feel free to add `TFAddLogSink(std::unique_ptr<TFLogSink>)` and/or `TFAddLogSink(std::shared_ptr<TFLogSink>)` in addition.  You can keep seperate `std::vector<std::unique_ptr<TFLogSink>>` and/or `std::vector<std::shared_ptr<TFLogSink>>` for `TFRemoveLogSink`.

In this way, we can support any lifetime signatures.

Reply:
Sounds good. I don't think having it both ways is a good idea. Managing different ownership models in the same registry is going to lead to some ugly and confusing code. I'm just going to go with your model with the user retaining ownership of the sinks. I did some thinking and I think at the end of the day, there are issues with either model of ownership, so I'll just stick to one and document it. Making things too clever introduces more problems than it solves.

I'll make the final changes some time today.